### PR TITLE
feat(fulltext_index): integrate full-text indexer with parquet reader

### LINKED ISSUE #4246: Tracking Issue: support full-text index
### What problem does the new feature solve?

Support full-text index.

### What does the feature do?

* [x] Creator for full-text index
* [x] Integrate with sst writer
* [x] CREATE statement support
* [ ] push down `matches` and apply the index

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

close #4246 

## What's changed and what's your intention?

* Add `FulltextIndexSearcher` and corresponding implementation `TantivyFulltextIndexSearcher`
* Integrate `FulltextIndexSearcher` into mito2's `FulltextIndexApplier`
* Enhance prune row groups logic, add `prune_row_groups_by_fulltext_index` based on `prune_row_groups_by_inverted_index`

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced full-text search functionality using Tantivy.
  - Added support for full-text index application and search in the scanning process.
  
- **Bug Fixes**
  - Improved error handling for full-text and inverted indexes.

- **Refactor**
  - Renamed various internal structs and methods for clarity, particularly around full-text and inverted index components.

- **Performance**
  - Enhanced metrics tracking for operations related to index application.

- **Tests**
  - Added comprehensive tests for full-text search scenarios.

### CODE REVIEW

File: src/puffin/src/puffin_manager/fs_puffin_manager/reader.rs [hunk 1]
```diff
@@ -161,13 +163,24 @@ where
...
+                    Ok(size)
+                });
+                tasks.push(task);
             }
 
+            let size = futures::future::try_join_all(tasks.into_iter())
+                .await
+                .into_iter()
+                .flatten()
+                .try_fold(0, |x, y| Ok(x + y?))?;
```
Comment (evenyag) [inline src/puffin/src/puffin_manager/fs_puffin_manager/reader.rs:182]: non-blocking: Can we use sum?
```suggestion
                .await?
                .into_iter()
                .sum::<u64>();
```

**Comment by evenyag** [comment]
Sqlness failed
```
thread 'grpc-handlers' panicked at src/mito2/src/sst/parquet/reader.rs:539:13:
attempt to subtract with overflow
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f9b16149208c8a8a349c32813312716f6603eb6f/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/f9b16149208c8a8a349c32813312716f6603eb6f/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_const::panic_const_sub_overflow
             at /rustc/f9b16149208c8a8a349c32813312716f6603eb6f/library/core/src/panicking.rs:175:21
   3: mito2::sst::parquet::reader::ParquetReaderBuilder::prune_row_groups_by_inverted_index::{{closure}}
             at /home/runner/work/greptimedb/greptimedb/src/mito2/src/sst/parquet/reader.rs:539:13
   4: mito2::sst::parquet::reader::ParquetReaderBuilder::row_groups_to_read::{{closure}}
             at /home/runner/work/greptimedb/greptimedb/src/mito2/src/sst/parquet/reader.rs:364:14
   5: mito2::sst::parquet::reader::ParquetReaderBuilder::build_reader_input::{{closure}}
             at /home/runner/work/greptimedb/greptimedb/src/mito2/src/sst/parquet/reader.rs:224:14
   6: mito2::read::scan_region::ScanInput::prune_file_ranges::{{closure}}
             at /home/runner/work/greptimedb/greptimedb/src/mito2/src/read/scan_region.rs:622:18
```

**Comment by evenyag** [comment]
Please fix the clippy warning
https://github.com/GreptimeTeam/greptimedb/actions/runs/9918929327/job/27404187601?pr=4348

**Comment by zhongzc** [comment]
Unit tests has been completed. Thank you for your hard work to review  @evenyag @v0y4g3r @waynexia @fengjiachun

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 2]
```diff
@@ -293,5 +293,273 @@ impl SingleCreator {
...
+    }
+
+    fn new_batch(
+        rows: impl Iterator<
+            Item = (
+                Option<impl AsRef<str>>, // text_english_case_sensitive
+                Option<impl AsRef<str>>, // text_english_case_insensitive
+                Option<impl AsRef<str>>, // text_chinese
+            ),
+        >,
```
Comment (evenyag) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:395]: Can we take a slice for simplicity?
```suggestion
        rows: &[
            (
                Option<&str>, // text_english_case_sensitive
                Option<&str>, // text_english_case_insensitive
                Option<&str>, // text_chinese
            ),
        ],
```

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 3]
```diff
@@ -293,5 +293,273 @@ impl SingleCreator {
...
+
+    async fn build_applier_factory(
+        prefix: &str,
+        rows: impl Iterator<
+            Item = (
+                Option<impl AsRef<str>>, // text_english_case_sensitive
+                Option<impl AsRef<str>>, // text_english_case_insensitive
+                Option<impl AsRef<str>>, // text_chinese
+            ),
+        >,
```
Comment (evenyag) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:456]:
```suggestion
        rows: [
            (
                Option<&str>, // text_english_case_sensitive
                Option<&str>, // text_english_case_insensitive
                Option<&str>, // text_chinese
            ),
        ],
```

**Review by evenyag — APPROVED** [review]

File: src/mito2/src/sst/index/fulltext_index/applier.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    /// Applies the queries to the fulltext index of the specified SST file.
+    pub async fn apply(&self, file_id: FileId) -> Result<BTreeSet<RowId>> {
+        let _timer = INDEX_APPLY_ELAPSED
+            .with_label_values(&[TYPE_FULLTEXT_INDEX])
+            .start_timer();
+
+        let mut inited = false;
+        let mut row_ids = BTreeSet::new();
+
+        for (column_id, query) in &self.queries {
```
Comment (fengjiachun) [inline src/mito2/src/sst/index/fulltext_index/applier.rs:76]: Have we considered executing multiple queries in parallel?

**Review by fengjiachun — APPROVED** [review]

File: src/mito2/src/sst/index/fulltext_index/applier.rs (hunk 4 above)
Comment (evenyag) [reply src/mito2/src/sst/index/fulltext_index/applier.rs:76]: Normally there should be only one match function so I think we can postpone it.

**Pull request merged**